qualtran/bloqs/state_preparation/state_preparation_alias_sampling.py

```diff
@@ -235,6 +235,219 @@ def build_call_graph(self, ssa: 'SympySymbolAllocator') -> Set['BloqCountT']:
+    )
+    index: Union[Shaped, NDArray[np.int_]] = attrs.field(eq=_data_or_shape_to_tuple)
+    alt: Union[Shaped, NDArray[np.int_]] = attrs.field(eq=_data_or_shape_to_tuple)
+    keep: Union[Shaped, NDArray[np.int_]] = attrs.field(eq=_data_or_shape_to_tuple)
+    mu: SymbolicInt
+    sum_of_lcu_coeffs: SymbolicFloat
+
+    @classmethod
+    def from_lcu_probs(
+        cls,
+        lcu_probabilities: Sequence[float],
+        *,
+        probability_epsilon: float = 1.0e-5,
+        nonzero_epsilon: float = 1e-6,
+    ) -> 'SparseStatePreparationAliasSampling':
+        """Factory to construct the state preparation gate for a given set of LCU coefficients.
+
+        Args:
+            lcu_probabilities: The LCU coefficients.
+            probability_epsilon: The desired accuracy to represent each probability
+                (which sets mu size and keep/alt integers).
+                See `qualtran.linalg.lcu_util.preprocess_lcu_coefficients_for_reversible_sampling`
+                for more information.
+            nonzero_epsilon: minimum value for a probability entry to be considered non-zero.
+        """
+        alt, keep, mu = preprocess_lcu_coefficients_for_reversible_sampling(
+            lcu_coefficients=lcu_probabilities, epsilon=probability_epsilon
+        )
+        N = len(lcu_probabilities)
+
+        is_nonzero = ~np.isclose(lcu_probabilities, 0, atol=nonzero_epsilon)
+        index = np.arange(N)[is_nonzero]
+        alt = np.array(alt)[is_nonzero]
+        keep = np.array(keep)[is_nonzero]
+
+        return cls(
+            selection_registers=Register('selection', BoundedQUInt((N - 1).bit_length(), N)),
+            index=index,
+            alt=alt,
+            keep=keep,
+            mu=mu,
+            sum_of_lcu_coeffs=sum(abs(x) for x in lcu_probabilities),
+        )
+
+    @classmethod
+    def from_n_coeff(
+        cls,
+        n_coeff: SymbolicInt,
+        n_nonzero_coeff: SymbolicInt,
+        sum_of_lcu_coeffs: SymbolicFloat,
+        *,
+        probability_epsilon: SymbolicFloat = 1.0e-5,
+    ) -> 'SparseStatePreparationAliasSampling':
+        """Factory to construct the state preparation gate for symbolic number of LCU coefficients.
+
+        Args:
+            n_coeff: Symbolic number of LCU coefficients in the prepared state.
+            n_nonzero_coeff: Symbolic number of non-zero LCU coefficients in the prepared state.
+            sum_of_lcu_coeffs: Sum of absolute values of coefficients of the prepared state.
+            probability_epsilon: The desired accuracy to represent each probability
+                (which sets mu size and keep/alt integers).
+                See `qualtran.linalg.lcu_util.preprocess_lcu_coefficients_for_reversible_sampling`
+                for more information.
+        """
+        mu = sub_bit_prec_from_epsilon(n_coeff, probability_epsilon)
+        selection_bitsize = bit_length(n_coeff - 1)
```

Comment:
Are you sure this should be `n_coeff` and not `n_nonzero_coeff` ?

Reply:
It is `n_coeff`. The selection register actually stays the same as the true LCU selection register (i.e. over `[0, L)`). I use an extra junk register called `sparse_index` which iterates over `[0, d)` (d = sparsity), which is used to prepare the state.